📖 Concepts and Terms (#37)
jupyter-book/myst-enhancement-proposals: meps/mep-0005.md

---
label: MEP0005
tags:
  - Draft
title: Concepts and Terms
date: 2025/12/10
authors:
  - stefanv

> short_title: "Concept & Terms"

Reviewer: ```suggestion
short_title: "Concepts & Terms"
```

description: Develops a standard naming convention for the various parts of the MyST processing pipeline.
data:

>   discussion: ''

Reviewer 2: Could we add a link to the Discord discussion? or copy-paste more of it to [the discussion here](https://github.com/jupyter-book/mystmd/discussions/2567) and link that? I feel like that's a really useful historical artifact for others to learn more.

---

## Summary

The MyST processing pipeline comprises various processing steps. This MEP tries to find a common language to describe them.

The suggested terms are:

...

## Context

> The MyST engine has various processing stages and, over time and as it developed, some confusing naming conventions came about.

Reviewer: This might be a style change, rather than substance... any excuse to write the word "organic" haha
```suggestion
Like all human labours, the MyST engine and the wider ecosystem that encompasses it has grown over time in its complexity. Following this organic trajectory, some confusing naming conventions and terminology have arisen.
```

For example, in MyST parlance "theme" and "template" mean roughly the same, but to the outside world "theme" has a strong aesthetic connection.
This document aims not only to capture the current status quo, but also to suggest a clarified naming scheme that can be used to standardize code and documentation.

## Pipeline steps

> The MyST pipeline takes an input MyST markdown file and proceeds through the following steps:

Reviewer: ```suggestion
The MyST pipeline takes an input MyST Markdown file and proceeds through the following steps:
```

:::{glossary}

> Parse

Reviewer 2: I've found that it is helpful to describe "inputs" and "outputs" in pipeline steps like this. E.g.:

1. Parse - input is source files (`.md`, `.ipynb`, etc), output is MyST AST
2. Transform - input is MyST AST, output is MyST AST
3. Render - input is MyST AST, output is (??) `[components / output nodes / chris isn't sure what to call this]`
4. Compile - input is `[whatever we call it above]`, output is `[final output format of the renderer]`
5. Start - For HTML outputs, spins up a webserver to view them in browser

: Take MyST markdown and produce a [myst-spec](https://mystmd.org/spec) compliant Abstract Syntax Tree (AST).
  You can see examples of these representations in the [MyST sandbox](https://mystmd.org/sandbox).

Transform
: At this stage, cross-reference resolution and other post-processing steps are performed, altering the AST.
  The various operators here are called `transformers` and include [... examples ...].

Render
: Convert one or more AST nodes into the output format.
  E.g., an HTML renderer would take an AST node and turn it into an HTML tag.

Compile
: Produce an output format from several inputs, including likely the AST.

Start
: Some output formats, such as web servers, need to be started after being compiled.
:::

> 

Reviewer: I'm tempted to advise against describing `start` here, especially as a consequence of `compile`. I suspect that the future involves the current lapreprint-template becoming more like a web-theme, where the symmetry is more obvious. It might be that we move `html` rendering _into_ the MyST theme given how coupled it is to myst-theme internals, so that the symmetry between web themes and static exporters is trivial.

## Pipeline actors

There are various actors in the pipeline that performs the above steps:

:::{glossary}

> myst-cli
> : The `myst` binary invoked by the user. Responsible for orchestrating the entire 
>   compilation pipeline.

Reviewer: ```suggestion
MyST Document Engine
: The `mystmd` binary invoked by the user. Responsible for orchestrating the entire 
  compilation pipeline.
```

Transformer
: An AST pre-processor.

Extension

> : A Javascript plugin that can provide roles/directives, inject AST, and/or be a transformer.

Reviewer: ### 
```suggestion
: A JavaScript plugin that can add new functionality to the MyST Document Engine. Presently, a plugin may provide roles/directives, inject AST, and/or be a transformer.
```

> template

Reviewer: Is the casing of these terms important here? Or should this be capitalised?

> : As per the dictionary definition[^template-dictionary-entry], a

Reviewer: I raised this on the Discord thread somewhere — I'm not sure whether the _conceptual_ term `template` is a helpful way to talk about themes and static templates (and I think we're using the term "template" here to mean the concept, not a repo with `template.yml`). If we drop the fact that we've used the word `template` in a technical sense, I think right now web themes and static templates are fairly conceptually distinct.

In a sentence, now:

> MyST Themes are applications that render MyST documents as websites. This is nominally done by rendering AST into e.g. React components.
> 
> MyST Static Templates are compilation recipes for weaving content (such as Typst or LaTeX) that is rendered by the MyST Document Engine from the MyST AST.

In the future, as per above, I think this might become:

> MyST Themes are applications that render MyST documents as websites. This is nominally done by rendering AST into e.g. React components.
> 
> MyST Static Templates are applications that render MyST documents into static formats like PDFs, or source-code to generate such documents such as LaTeX files.

  template provides a pattern through which outputs are produced.  A
  template can be static (e.g., a LaTeX template from which a PDF is produced),
  or dynamic (a piece of software that produces a web app).[^theme-vs-template]
:::

[^template-dictionary-entry]: "A generic model or pattern from which other objects are based or derived." — The American Heritage, Dictionary of the English Language, 5th Edition
[^theme-vs-template]: Templates used to be known as themes, but this caused some confusion (since themes are typically related to layout and styling). In this MEP, we simply use `template` throughout.

## Actor scopes

It is important to note the scopes of (i.e., the pipeline steps performed by) of the above actors. Specifically:

1. The `myst-cli` is the orchestrator in control of the entire production pipeline.
   For convenience, it can also be used to *start* certain output formats (such as web servers).
   E.g., if the output is a web application, it can start it up.

> 2. Templates can vary in their complexity, and in how much of the compilation pipeline they control.

Reviewer: Hmm, now I'm not sure if we're using Template to in-fact refer to `template.yml` projects?

   For example, consider two output formats:

   - **PDF from LaTeX:** A *static LaTeX template* combines LaTeX
     snippets.  The `myst-cli` *renders* these snippets, and
     also invokes PDFLaTeX (with a *template* as *input*) to
     *compile* the template into a PDF.

>    - **Web application:** MyST's default web output is a *web

Reviewer: I'm not sure whether we need to introduce compiler in the contexts of MyST Themes — in theory the `mystmd` binary starts the `myst-theme` application, which consumes the nearly-raw AST. Maybe we can refer to `mystmd` preparing `_build/site` as a compilation step, but is that the intention?

     application*. The web application is *compiled*, by `myst-cli`,
     from a *dynamic template* (a code package). As a convenience,
     `myst-cli` will also launch the web app.

     The web application *itself* is a *renderer*.  It consumes AST
     (from a content server or `.json` file), *renders* the necessary
     React components, and then *renders* server pages,
     browsable by the user.

Actor scopes are currently vaguely defined. For example, when building
a new HTML theme, is `myst-cli` responsible for generating HTML
components, or does that happen inside the template itself? These
boundaries will be explored further over time.

## Example language

- "When upgrading the myst-spec, *renderers* are the only actors affected."

## Proposal

We propose to adopt the language from the glossaries above throughout
the code, as well as in documentation.

> For example, we would rename the current `myst-theme` to `myst-webapp-template`,

Reviewer 3: There is also the `@myst-theme` which is the NPM organization that these packages are published in. That also needs to be considered...!

Reviewer: Despite Rowans' fair point, whatever name we choose, I like the idea of dropping Theme. The baggage is severe.

## Pipeline steps

The MyST pipeline takes an input MyST markdown file and proceeds through the following steps:

:::{glossary}
Parse
: Take MyST markdown and produce a [myst-spec](https://mystmd.org/spec) compliant Abstract Syntax Tree (AST).
  You can see examples of these representations in the [MyST sandbox](https://mystmd.org/sandbox).

Transform
: At this stage, cross-reference resolution and other post-processing steps are performed, altering the AST.
  The various operators here are called `transformers` and include [... examples ...].

Render
: Convert one or more AST nodes into the output format.
  E.g., an HTML renderer would take an AST node and turn it into an HTML tag.

Compile
: Produce an output format from several inputs, including likely the AST.

Execute
: Some formats need to be launched after being compiled.  E.g., web
  servers. Many output formats, like PDF, do not require execution.
:::

## Pipeline actors

There are various actors in the pipeline that performs the above steps:

:::{glossary}
myst-cli
: The `myst` binary invoked by the user. Responsible for orchestrating the entire 
  compilation pipeline.

Transformer
: An AST pre-processor.

Extension

> : A Javascript plugin that can inject AST and/or be a transformer.

Reviewer 3: ```suggestion
: A Javascript plugin that can provide roles/directives, inject AST, and/or be a transformer.
```
